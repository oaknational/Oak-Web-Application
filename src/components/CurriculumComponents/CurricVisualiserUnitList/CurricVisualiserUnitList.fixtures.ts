import { createFilter } from "@/fixtures/curriculum/filters";
import { createUnit } from "@/fixtures/curriculum/unit";
import { createYearData } from "@/fixtures/curriculum/yearData";

export const basicFixtures = {
  units: [
    createUnit({ slug: "one" }),
    createUnit({ slug: "two" }),
    createUnit({ slug: "three" }),
    createUnit({ slug: "four" }),
  ],
  yearData: {
    "7": createYearData({}),
  },
  year: "1",
  filters: createFilter({}),
  basePath: "/",
};
