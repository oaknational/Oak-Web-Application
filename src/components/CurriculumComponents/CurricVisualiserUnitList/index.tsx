import { join } from "path";

import { OakFlex, OakP } from "@oaknational/oak-components";
import styled from "styled-components";
import { useSearchParams } from "next/navigation";

import CurricUnitCard from "../CurricUnitCard";

import { isHighlightedUnit } from "@/utils/curriculum/filtering";
import {
  CurriculumFilters,
  Thread,
  Unit,
  YearData,
} from "@/utils/curriculum/types";
import { getSubjectCategoryMessage } from "@/utils/curriculum/formatting";
import useAnalytics from "@/context/Analytics/useAnalytics";
import useAnalyticsPageProps from "@/hooks/useAnalyticsPageProps";
import { buildUnitOverviewAccessedAnalytics } from "@/utils/curriculum/analytics";

const UnitList = styled("ol")`
  margin: 0;
  list-style: none;
  padding: 0;
  display: flex;
  flex-wrap: wrap;
  gap: 16px;
`;

const UnitListItem = styled("li")`
  margin: 0;
  liststyle: none;
  padding: 0;
  display: flex;
  width: 240px;
  flex-grow: 1;
  position: relative;
`;

type CurricVisualiserUnitListProps = {
  units: Unit[];
  filters: CurriculumFilters;
  year: string;
  yearData: YearData;
  basePath: string;
  selectedThread?: Thread;
};
export function CurricVisualiserUnitList({
  units,
  yearData,
  year,
  filters,
  basePath,
  selectedThread,
}: CurricVisualiserUnitListProps) {
  const { track } = useAnalytics();
  const { analyticsUseCase } = useAnalyticsPageProps();
  const searchParams = useSearchParams();

  const onClick = (unit: Unit, isHighlighted: boolean) => {
    track.unitOverviewAccessed(
      buildUnitOverviewAccessedAnalytics({
        unit,
        isHighlighted,
        componentType: "unit_info_button",
        selectedThread,
        analyticsUseCase,
      }),
    );
  };

  function getItems(unit: Unit, index: number) {
    const isHighlighted = isHighlightedUnit(unit, filters.threads);
    const searchParamsStr = searchParams?.toString() ?? "";
    const unitUrl =
      join(basePath, unit.slug) +
      `${!searchParamsStr ? "" : `?${searchParamsStr}`}`;

    return (
      <UnitListItem key={`${unit.slug}-${index}`}>
        <CurricUnitCard
          unit={unit}
          key={unit.slug + index}
          index={index}
          isHighlighted={isHighlighted}
          href={unitUrl}
          onClick={() => onClick(unit, isHighlighted)}
        />
      </UnitListItem>
    );
  }

  return (
    <OakFlex
      $flexWrap={"wrap"}
      $pt="inner-padding-s"
      data-testid="unit-cards"
      $gap={"all-spacing-4"}
      // TODO: Remove hack
      style={{
        marginBottom: "-1rem",
      }}
    >
      <UnitList role="list">
        {units.length < 1 && (
          <OakP>
            {getSubjectCategoryMessage(
              yearData,
              year,
              filters.subjectCategories,
            )}
          </OakP>
        )}
        {units.map(getItems)}
        {/* Empty tiles for correct flex wrapping */}
        {Array(3)
          .fill(true)
          .map((item, index) => {
            return (
              <OakFlex
                key={`unit-list-item-${item}-${index}`}
                $width={"all-spacing-19"}
                $flexGrow={1}
                $position={"relative"}
              />
            );
          })}
      </UnitList>
    </OakFlex>
  );
}
