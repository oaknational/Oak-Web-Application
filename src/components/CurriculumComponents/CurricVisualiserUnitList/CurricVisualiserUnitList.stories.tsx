import { OakThemeProvider, oakDefaultTheme } from "@oaknational/oak-components";
import { Meta, StoryObj } from "@storybook/react";

import { basicFixtures } from "./CurricVisualiserUnitList.fixtures";

import { CurricVisualiserUnitList as Component } from ".";

const meta: Meta<typeof Component> = {
  component: Component,
  argTypes: {},
};

export default meta;
type Story = StoryObj<typeof Component>;

export const CurricVisualiserUnitList: Story = {
  args: {
    ...basicFixtures,
  },
  render: function Render(args) {
    return (
      <OakThemeProvider theme={oakDefaultTheme}>
        <Component {...args} />
      </OakThemeProvider>
    );
  },
};
