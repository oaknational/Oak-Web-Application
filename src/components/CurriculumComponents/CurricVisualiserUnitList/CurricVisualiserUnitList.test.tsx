import { basicFixtures } from "./CurricVisualiserUnitList.fixtures";

import { CurricVisualiserUnitList } from ".";

import renderWithTheme from "@/__tests__/__helpers__/renderWithTheme";

describe("CurricVisualiserUnitList", () => {
  it("should display a list of units", async () => {
    const { baseElement, findAllByRole } = renderWithTheme(
      <CurricVisualiserUnitList {...basicFixtures} />,
    );

    const allLinks = await findAllByRole("link");
    expect(allLinks.length).toEqual(4);
    expect(allLinks[0]?.textContent).toEqual(expect.stringContaining("One"));
    expect(allLinks[1]?.textContent).toEqual(expect.stringContaining("Two"));
    expect(allLinks[2]?.textContent).toEqual(expect.stringContaining("Three"));
    expect(allLinks[3]?.textContent).toEqual(expect.stringContaining("Four"));
    expect(baseElement).toMatchSnapshot();
  });
});
