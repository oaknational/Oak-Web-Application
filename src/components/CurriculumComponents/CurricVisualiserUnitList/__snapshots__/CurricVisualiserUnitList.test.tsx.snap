// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurricVisualiserUnitList should display a list of units 1`] = `
<body>
  <div>
    <div
      class="sc-fqkvVR sc-gsFSXq gOQIrG kMhoXH"
      data-testid="unit-cards"
      style="margin-bottom: -1rem;"
    >
      <ol
        class="sc-a964d3a7-0 csZIsZ"
        role="list"
      >
        <li
          class="sc-a964d3a7-1 hFJsag"
        >
          <div
            class="sc-fqkvVR sc-1c55d370-0 jdSkEU iHxKuU"
          >
            <a
              class="sc-79c3f222-0 bginFa"
              data-testid="unit-info-link"
              href="/one"
            >
              <div
                class="sc-fqkvVR sc-gsFSXq hjukdb bOTBNE"
                data-testid="unit-card"
              >
                <div
                  class="sc-fqkvVR fgBWrW"
                >
                  <div
                    class="sc-9d6b6245-0 cPzdPh pa11y-ignore"
                  >
                    <span
                      aria-hidden="true"
                    >
                      1
                    </span>
                  </div>
                  <span
                    class="sc-kpDqfm fsPjOh"
                  >
                    One
                  </span>
                </div>
                <div
                  class="sc-fqkvVR sc-gsFSXq julalA liZwDP"
                >
                  <div
                    class="sc-fqkvVR sc-gsFSXq fgBWrW gJyPoc"
                  >
                    <div
                      class="sc-fqkvVR sc-eldPxv kxMTGH dHEUFw"
                    >
                      Unit info
                    </div>
                    <div
                      class="sc-fqkvVR kbnxKi"
                    >
                      <img
                        alt=""
                        class="sc-dcJsrY beWtwI"
                        data-nimg="fill"
                        decoding="async"
                        loading="lazy"
                        src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1707752509/icons/vk9xxxhnsltsickom6q9.svg"
                        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </a>
          </div>
        </li>
        <li
          class="sc-a964d3a7-1 hFJsag"
        >
          <div
            class="sc-fqkvVR sc-1c55d370-0 jdSkEU iHxKuU"
          >
            <a
              class="sc-79c3f222-0 bginFa"
              data-testid="unit-info-link"
              href="/two"
            >
              <div
                class="sc-fqkvVR sc-gsFSXq hjukdb bOTBNE"
                data-testid="unit-card"
              >
                <div
                  class="sc-fqkvVR fgBWrW"
                >
                  <div
                    class="sc-9d6b6245-0 cPzdPh pa11y-ignore"
                  >
                    <span
                      aria-hidden="true"
                    >
                      2
                    </span>
                  </div>
                  <span
                    class="sc-kpDqfm fsPjOh"
                  >
                    Two
                  </span>
                </div>
                <div
                  class="sc-fqkvVR sc-gsFSXq julalA liZwDP"
                >
                  <div
                    class="sc-fqkvVR sc-gsFSXq fgBWrW gJyPoc"
                  >
                    <div
                      class="sc-fqkvVR sc-eldPxv kxMTGH dHEUFw"
                    >
                      Unit info
                    </div>
                    <div
                      class="sc-fqkvVR kbnxKi"
                    >
                      <img
                        alt=""
                        class="sc-dcJsrY beWtwI"
                        data-nimg="fill"
                        decoding="async"
                        loading="lazy"
                        src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1707752509/icons/vk9xxxhnsltsickom6q9.svg"
                        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </a>
          </div>
        </li>
        <li
          class="sc-a964d3a7-1 hFJsag"
        >
          <div
            class="sc-fqkvVR sc-1c55d370-0 jdSkEU iHxKuU"
          >
            <a
              class="sc-79c3f222-0 bginFa"
              data-testid="unit-info-link"
              href="/three"
            >
              <div
                class="sc-fqkvVR sc-gsFSXq hjukdb bOTBNE"
                data-testid="unit-card"
              >
                <div
                  class="sc-fqkvVR fgBWrW"
                >
                  <div
                    class="sc-9d6b6245-0 cPzdPh pa11y-ignore"
                  >
                    <span
                      aria-hidden="true"
                    >
                      3
                    </span>
                  </div>
                  <span
                    class="sc-kpDqfm fsPjOh"
                  >
                    Three
                  </span>
                </div>
                <div
                  class="sc-fqkvVR sc-gsFSXq julalA liZwDP"
                >
                  <div
                    class="sc-fqkvVR sc-gsFSXq fgBWrW gJyPoc"
                  >
                    <div
                      class="sc-fqkvVR sc-eldPxv kxMTGH dHEUFw"
                    >
                      Unit info
                    </div>
                    <div
                      class="sc-fqkvVR kbnxKi"
                    >
                      <img
                        alt=""
                        class="sc-dcJsrY beWtwI"
                        data-nimg="fill"
                        decoding="async"
                        loading="lazy"
                        src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1707752509/icons/vk9xxxhnsltsickom6q9.svg"
                        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </a>
          </div>
        </li>
        <li
          class="sc-a964d3a7-1 hFJsag"
        >
          <div
            class="sc-fqkvVR sc-1c55d370-0 jdSkEU iHxKuU"
          >
            <a
              class="sc-79c3f222-0 bginFa"
              data-testid="unit-info-link"
              href="/four"
            >
              <div
                class="sc-fqkvVR sc-gsFSXq hjukdb bOTBNE"
                data-testid="unit-card"
              >
                <div
                  class="sc-fqkvVR fgBWrW"
                >
                  <div
                    class="sc-9d6b6245-0 cPzdPh pa11y-ignore"
                  >
                    <span
                      aria-hidden="true"
                    >
                      4
                    </span>
                  </div>
                  <span
                    class="sc-kpDqfm fsPjOh"
                  >
                    Four
                  </span>
                </div>
                <div
                  class="sc-fqkvVR sc-gsFSXq julalA liZwDP"
                >
                  <div
                    class="sc-fqkvVR sc-gsFSXq fgBWrW gJyPoc"
                  >
                    <div
                      class="sc-fqkvVR sc-eldPxv kxMTGH dHEUFw"
                    >
                      Unit info
                    </div>
                    <div
                      class="sc-fqkvVR kbnxKi"
                    >
                      <img
                        alt=""
                        class="sc-dcJsrY beWtwI"
                        data-nimg="fill"
                        decoding="async"
                        loading="lazy"
                        src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1707752509/icons/vk9xxxhnsltsickom6q9.svg"
                        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </a>
          </div>
        </li>
        <div
          class="sc-fqkvVR sc-gsFSXq iASqjD hKscQn"
        />
        <div
          class="sc-fqkvVR sc-gsFSXq iASqjD hKscQn"
        />
        <div
          class="sc-fqkvVR sc-gsFSXq iASqjD hKscQn"
        />
      </ol>
    </div>
  </div>
</body>
`;
